### PR TITLE
Test perturbed cost

### PR DESCRIPTION
**Pull request by BatyLeo**
Test the *learn by experience* `PerturbedCost` implementation :
- Tests on argmax and ranking, unified with `simple_optimizers` tests.
- Fix some bugs in `simple_optimizers` tests.
- Add more metric tests in `simple_optimizers` tests.

Note: accuracy check in `PerturbedCost` test for ranking is currently set to `@test_broken`.

### CODE REVIEW

File: test/simple_optimizers.jl [hunk 1]
```diff
@@ -3,13 +3,14 @@
 N = 1000
 dim_x = 10
 dim_y = 8
-σ = 0.01
+σ = 0.0
```
Comment (gdalle) [inline test/simple_optimizers.jl:6]: Why didn't you keep sigma > 0 ?

File: test/simple_optimizers.jl [hunk 2]
```diff
@@ -3,13 +3,14 @@
...
 epochs = 200
 
 ## True model
 
 true_model = Dense(dim_x, dim_y; bias=false)
-true_optimizer = Dict("argmax" => one_hot_argmax, "ranking" => ranking)
+true_optimizer = Dict("argmax" => argmax_optimizer, "ranking" => ranking)
+cost(y; instance) = -true_model(instance)' * y
```
Comment (gdalle) [inline test/simple_optimizers.jl:13]: Would it be more natural in terms of code to maximize the negative cost and call it benefit or something?

File: test/simple_optimizers.jl [hunk 3]
```diff
@@ -70,10 +71,16 @@ for setting in ["argmax", "ranking"], target in ["y", "θ", "none"]
 
         if target == "y"
             data = zip(X_train, Y_train)
+            data_test = zip(X_test, Y_test)
         elseif target == "θ"
             data = zip(X_train, thetas_train)
+            data_test = zip(X_test, thetas_test)
+        elseif target == "(θ,y)"
```
Comment (gdalle) [inline test/simple_optimizers.jl:78]: Nice catch

File: test/utils.jl [hunk 4]
```diff
@@ -3,7 +3,11 @@
 mape(x1, x2) = 100 * mean(abs.((x1 .- x2) ./ x1))
 normalized_mape(x1, x2) = mape(x1 / norm(x1), x2 / norm(x2))
 
-function ranking(θ::AbstractVector; rev::Bool=false)
+function argmax_optimizer(θ::AbstractVector; instance::AbstractVector=Float64[])
```
Comment (gdalle) [inline test/utils.jl:6]: Why does the instance have to be an AbstractVector? I would do `instance=nothing` instead

**Review by gdalle — CHANGES_REQUESTED** [review]
Thanks for the additional tests! See my comments for a few (potential) improvements

File: test/simple_optimizers.jl (hunk 1 above)
Comment (BatyLeo) [reply test/simple_optimizers.jl:6]: When sigma > 0, the value of the `objective_gap` metric (only computed when `target == "none"`) is not entirely correct : the optimal `y` in the *learning by experience* case is `optimizer(true_model(x))`, which is noisy when `sigma > 0`. A solution to keep `sigma > 0` is to compute an `unperturbed_Y` during the `generate_dataset` and use it to compute the `objective_gap`.

File: test/simple_optimizers.jl (hunk 2 above)
Comment (BatyLeo) [reply test/simple_optimizers.jl:13]: If we use `PerturbedCost` as a loss, its `cost` argument is minimized, so we have to use this formulation here I think

File: test/utils.jl (hunk 4 above)
Comment (BatyLeo) [reply test/utils.jl:6]: I agree

File: test/utils.jl [hunk 5]
```diff
@@ -21,9 +25,11 @@ function generate_dataset(
...
 )
     X = [randn(dim_x) for n in 1:N]
-    thetas = [model(X[n]) + σ * randn(dim_y) for n in 1:N]
-    Y = [optimizer(θ) for θ in thetas]
-    return (X=X, thetas=thetas, Y=Y)
+    thetas = [model(x) for x in X]
+    noiseless_Y = [optimizer(θ) for θ in thetas]
+    noisy_thetas = [θ + σ * randn(dim_y) for θ in thetas]
+    noisy_Y = [optimizer(θ) for θ in noisy_thetas]
+    return (X=X, thetas=thetas, noisy_Y=noisy_Y, noiseless_Y=noiseless_Y)
```
Comment (gdalle) [inline test/utils.jl:32]: Not sure we need to generate both: I actually meant to use `σ = 0` to generate the traning set and `σ > 0` to generate the test set. Is that possible?

**Review by gdalle — DISMISSED** [review]

File: test/utils.jl (hunk 5 above)
Comment (gdalle) [reply test/utils.jl:32]: Solved by @BatyLeo - we actually need a noiseless Y_train to compute the objective gap

**Review by gdalle — APPROVED** [review]

**Pull request merged**